controllers/control_plane_init_locker_test.go

```diff
@@ -17,219 +17,207 @@ limitations under the License.
 	}{
 		{
-			name:          "error getting configmap",
-			getError:      errors.New("get error"),
-			expectRelease: false,
+			name:          "should release lock by deleting config map",
+			context:       context.Background(),
+			client:        &fakeClient{},
+			shouldRelease: true,
 		},
 		{
-			name:          "configmap not found",
-			getError:      apierrors.NewNotFound(schema.GroupResource{Group: "", Resource: "configmaps"}, "uid1-configmap"),
-			expectRelease: true,
+			name:    "should not release lock if cant delete config map",
+			context: context.Background(),
+			client: &fakeClient{
+				deleteError: errors.New("delete error"),
+			},
+			shouldRelease: false,
 		},
 		{
-			name:          "delete succeeds",
-			expectRelease: true,
+			name:    "should release lock if config map does not exist",
+			context: context.Background(),
+			client: &fakeClient{
+				getError: apierrors.NewNotFound(schema.GroupResource{Group: "", Resource: "configmaps"}, fmt.Sprintf("%s-controlplane", uid)),
+			},
+			shouldRelease: true,
 		},
 		{
-			name:          "delete fails",
-			deleteError:   errors.New("delete error"),
-			expectRelease: false,
+			name:    "should not release lock if error while getting config map",
+			context: context.Background(),
+			client: &fakeClient{
+				getError: errors.New("get error"),
+			},
+			shouldRelease: false,
 		},
 	}
 
 	for _, tc := range tests {
 		t.Run(tc.name, func(t *testing.T) {
 			l := &controlPlaneInitLocker{
-				log: log.Log,
-				configMapClient: &configMapsGetter{
-					configMap:   tc.configMap,
-					getError:    tc.getError,
-					deleteError: tc.deleteError,
-				},
+				log:    log.ZapLogger(true),
+				ctx:    context.Background(),
+				client: tc.client,
 			}
 
 			cluster := &clusterv2.Cluster{
 				ObjectMeta: metav1.ObjectMeta{
-					Namespace: "ns1",
-					Name:      "name1",
-					UID:       types.UID("uid1"),
+					Namespace: clusterNamespace,
+					Name:      clusterName,
+					UID:       uid,
 				},
 			}
 
 			released := l.Release(cluster)
-			if tc.expectRelease != released {
-				t.Errorf("expected %t, got %t", tc.expectRelease, released)
+			if released != tc.shouldRelease {
+				t.Fatalf("released was %v, but it should be %v\n", released, tc.shouldRelease)
 			}
+
 		})
 	}
 }
 
-type configMapsGetter struct {
-	configMap   *v1.ConfigMap
-	getError    error
-	createError error
-	deleteError error
+type fakeStatusWriter struct{}
+
+func (fsw *fakeStatusWriter) Update(ctx context.Context, obj runtime.Object, opts ...client.UpdateOption) error {
+	panic("not implemented")
 }
 
-func (c *configMapsGetter) ConfigMaps(namespace string) corev1client.ConfigMapInterface {
-	return &configMapClient{
-		configMap:   c.configMap,
-		getError:    c.getError,
-		createError: c.createError,
-		deleteError: c.deleteError,
-	}
+func (fsw *fakeStatusWriter) Patch(ctx context.Context, obj runtime.Object, patch client.Patch, opts ...client.PatchOption) error {
```

Comment:
the controller-runtime `client` package provides a fake client for us that we can use instead of this.

Reply:
I looked at the fake client that the `client` package provides. I wanted to ability to throw errors in the `Create`, `Delete` and `Get` methods as needed so as to better unit test the locker. I don't think the fake client in the `client` package has this ability.

Reply:
I'd second using the `fake` client, and only interact with errors that come from using it, instead of dealing with arbitrary ones.

Reply:
I explored using the `fake` client while writing these unit tests, however, some of the unit tests that need to be covered here are cases like "do not acquire the lock if cannot create config map". This means we need to replicate a case where the client's `Create` call fails. The `fake` client does not have a mechanism to set this up so as to test the case. There are similar cases in lock release as well. Hence I ended up creating this fake client in these unit tests.

We can use the `fake` client but that would mean we cannot test all the cases of the init locker.